File: experiments/source/cpp/weighted_dot.cpp
#include <loom/loom.hpp>

LOOM_ENTRY float weighted_dot(float weight, loom::Tensor1 x, loom::Tensor1 y) {
  auto contribution = [=](float xi, float yi) -> float { return weight * xi * yi; };
  return loom::Tensor1::reduce_sum(loom::Tensor1::map2(contribution, x, y));
}

File: experiments/source/python/weighted_dot.py
import loom


@loom.entry
def weighted_dot(weight: float, x: loom.Tensor1, y: loom.Tensor1) -> float:
    return loom.Tensor1.reduce_sum(loom.Tensor1.map2(lambda xi, yi: weight * xi * yi, x, y))
